### PR TITLE
Core: Extract listPartition methods from SparkTableUtil into DataUtil class in core module.

### PR DESCRIPTION
**Pull request by lcspinter**
SparkTableUtil introduced some utility methods (`public static List<DataFile> listPartition(Map<String, String> partitionKeys, String uri, String format, PartitionSpec spec, Configuration conf, MetricsConfig metricsConfig, NameMapping mapping`) which are used to generate `DataFile`s when importing files from existing Spark table into Iceberg table. These utility methods can be useful when migrating Hive table to Iceberg table, therefore it would make sense to extract them to the core module.

### CODE REVIEW

**Comment by lcspinter** [comment]
@marton-bod @pvary Could you please review this PR? Thank you

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+import org.apache.iceberg.PartitionField;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.hadoop.HadoopInputFile;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.orc.OrcMetrics;
+import org.apache.iceberg.parquet.ParquetUtil;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+
+public class DataUtil {
```
Comment (marton-bod) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:46]: Do we want to keep this as a generic utility class for all sorts of data helper operations? Or do we want this to contain only migration related helper methods, e.g. `TableMigrationUtil`?

**Comment by marton-bod** [comment]
@aokolnychyi @RussellSpitzer can you please take a look as well, since it touches the Spark/Data modules?

**Comment by aokolnychyi** [comment]
Let me take a look in a few hours.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java (hunk 1 above)
Comment (aokolnychyi) [reply data/src/main/java/org/apache/iceberg/data/DataUtil.java:46]: I agree with @marton-bod. The currently proposed name is not descriptive. I think `TableMigrationUtil` fits better.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+  private DataUtil() {
+  }
+
+  private static final PathFilter HIDDEN_PATH_FILTER =
+      p -> !p.getName().startsWith("_") && !p.getName().startsWith(".");
+
+    /**
+     * Returns the data files in a partition by listing the partition location.
+     *
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:56]: Looks like the overall formatting of the doc is off by one space.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+     * @param partitionKeys partition key, e.g., "a=1/b=2"
+     * @param uri partition location URI
+     * @param format partition format, avro, parquet or orc
+     * @param spec a partition spec
+     * @param conf a Hadoop conf
+     * @param metricsConfig a metrics conf
+     * @return a List of DataFile
+     */
+  public static List<DataFile> listPartition(Map<String, String> partitionKeys, String uri, String format,
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:68]: I think it should be either `partition` like before or `partitionKey` as Iceberg `PartitionKey` represents a struct of partition values.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+     * @param partitionKeys partition key, e.g., "a=1/b=2"
+     * @param uri partition location URI
+     * @param format partition format, avro, parquet or orc
+     * @param spec a partition spec
+     * @param conf a Hadoop conf
+     * @param metricsConfig a metrics conf
+     * @param mapping a name mapping
+     * @return a List of DataFile
+     */
+  public static List<DataFile> listPartition(Map<String, String> partitionKeys, String uri, String format,
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:91]: Same for `partitionKeys` here.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java (hunk 3 above)
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:68]: Actually, do we need this method without `NameMapping`? I think it has no usages now and we should definitely encourage a name mapping to be present during imports.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+     * @param conf a Hadoop conf
+     * @param metricsConfig a metrics conf
+     * @return a List of DataFile
+     */
+  public static List<DataFile> listPartition(Map<String, String> partitionKeys, String uri, String format,
+                                             PartitionSpec spec, Configuration conf, MetricsConfig metricsConfig) {
+    return listPartition(partitionKeys, uri, format, spec, conf, metricsConfig, null);
+  }
+
+    /**
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:73]: Doc formatting here too.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+  private static List<DataFile> listAvroPartition(Map<String, String> partitionPath, String partitionUri,
+                                                  PartitionSpec spec, Configuration conf) {
+    try {
+      Path partition = new Path(partitionUri);
+      FileSystem fs = partition.getFileSystem(conf);
+      return Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
+              .filter(FileStatus::isFile)
+              .map(stat -> {
+                  // Avro file statistics cannot be calculated without reading the file.
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:113]: This does not seem to match the existing logic. If we want to make this change, can we do that in a separate PR?
Also, won't it break the existing filtering logic?

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                  // Setting the rowCount to 0 is just a workaround so that the DataFiles.Builder.build() doesn't fail.
+                Metrics metrics = new Metrics(0L, null, null, null);
+                String partitionKey = spec.fields().stream()
+                          .map(PartitionField::name)
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:117]: Iceberg normally uses just 4 spaces for continued indentation. Can we format this class accordingly?

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+      return Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
+              .filter(FileStatus::isFile)
+              .map(stat -> {
+                Metrics metrics;
+                try {
+                  ParquetMetadata metadata = ParquetFileReader.readFooter(conf, stat);
+                  metrics = ParquetUtil.footerMetrics(metadata, Stream.empty(), metricsSpec, mapping);
+                } catch (IOException e) {
+                  throw new RuntimeException("Unable to read the footer of the parquet file: " +
+                              stat.getPath(), e);
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:151]: nit: formatting, should be 4 spaces for continued indentation

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                                                 MetricsConfig metricsSpec, NameMapping mapping) {
+    try {
+      Path partition = new Path(partitionUri);
+      FileSystem fs = partition.getFileSystem(conf);
+
+      return Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
+              .filter(FileStatus::isFile)
+              .map(stat -> {
+                Metrics metrics = OrcMetrics.fromInputFile(HadoopInputFile.fromPath(stat.getPath(), conf),
+                          metricsSpec, mapping);
```
Comment (aokolnychyi) [inline data/src/main/java/org/apache/iceberg/data/DataUtil.java:182]: formatting

File: spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java [hunk 10]
```diff
@@ -281,155 +270,10 @@ private SparkTableUtil() {
   public static List<DataFile> listPartition(SparkPartition partition, PartitionSpec spec,
```
Comment (aokolnychyi) [inline spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:270]: Shall we deprecate these methods?

**Comment by aokolnychyi** [comment]
@RussellSpitzer @flyrain, could you take a look too?

**Comment by RussellSpitzer** [comment]
Looks good to me aside from naming of the utility class and the formatting issues 👍

**Comment by flyrain** [comment]
Other than the naming and formatting issue, should it be in package org.apache.iceberg.util? Does it make sense to put them in class `PartitionUtil`?

**Comment by marton-bod** [comment]
My impression is that `PartitionUtil` is supposed to define operations on Iceberg partitions, whereas here we are dealing with listing Hive partitions so it might not be a good idea to mix the two. I'd prefer keeping this migration-related logic in this separate `TableMigrationUtil` class, especially if more methods will be added to it as migration efforts pick up pace.

File: data/src/main/java/org/apache/iceberg/data/DataUtil.java (hunk 6 above)
Comment (lcspinter) [reply data/src/main/java/org/apache/iceberg/data/DataUtil.java]: @aokolnychyi I changed back to the original value (-1L) and I will open a separate PR to address this issue. 
With -1 set as rowCount will result in [failure|https://github.com/apache/iceberg/blob/a79de571860a290f6e96ac562d616c9c6be2071e/core/src/main/java/org/apache/iceberg/DataFiles.java#L288] when calling DataFiles.Builder.build(). I checked the associated test suite `TestSparkTableUtil`, and it seems that the import of spark tables having data files in avro file format [was not tested|https://github.com/apache/iceberg/blob/a79de571860a290f6e96ac562d616c9c6be2071e/spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java#L139] at all.

**Comment by lcspinter** [comment]
I've uploaded a new PR with the requested formatting/renaming changes.

**Comment by RussellSpitzer** [comment]
Yeah, because of classloader issues I was only able two write tests for parquet. Both orc and Avro are set to ignore in the procedure/action tests

**Comment by marton-bod** [comment]
@aokolnychyi @RussellSpitzer Thanks for reviewing this!
As I see it now, the only outstanding thing is deprecating this method and otherwise good to go? https://github.com/apache/iceberg/pull/2494/commits/3d499cda445adbaf7c6ce7a529169a2d256e5159#diff-e9befa3e2e58983fca27df05bd7d62b673c327261522e3894bb5fc6d4800f17dR270
Thank you! cc @lcspinter

**Comment by lcspinter** [comment]
@RussellSpitzer @aokolnychyi I've marked the methods as deprecated. 
What is left if to solve this [issue] (https://github.com/apache/iceberg/pull/2494#discussion_r616607521) in a separate PR. With rowCount set to 0, the hive table migration works perfectly, but I'm not sure how will affect other use cases. Do you happen to know some other solution? Thanks

**Review by pvary — APPROVED** [review]

**Comment by RussellSpitzer** [comment]
@lcspinter I have no problem keeping the behavior of with the -1 as is for now even if it's broken. We should fix it when we get real working avro tests, which we probably can do in the integration suite I setup. It shouldn't have the same classloader baggage our normal test classpath has.

**Comment by RussellSpitzer** [comment]
I think everyone's given their +1's so i'm going to merge this. Thanks for the contribution @lcspinter !
